spring-integration-http/src/main/java/org/springframework/integration/http/outbound/AbstractHttpRequestExecutingMessageHandler.java

```diff
@@ -0,0 +1,539 @@
+		this.transferCookies = transferCookies;
+	}
+
+	@Override
+	protected void doInit() {
+		this.evaluationContext = ExpressionUtils.createStandardEvaluationContext(this.getBeanFactory());
+	}
+
+	@Override
+	protected Object handleRequestMessage(Message<?> requestMessage) {
+		Object uri = this.uriExpression.getValue(this.evaluationContext, requestMessage);
+		Assert.state(uri instanceof String || uri instanceof URI,
+				"'uriExpression' evaluation must result in a 'String' or 'URI' instance, not: "
+						+ (uri == null ? "null" : uri.getClass()));
+		URI realUri = null;
+		try {
+			HttpMethod httpMethod = this.determineHttpMethod(requestMessage);
+
+			if (!this.shouldIncludeRequestBody(httpMethod) && this.extractPayloadExplicitlySet) {
+				if (logger.isWarnEnabled()) {
+					logger.warn("The 'extractPayload' attribute has no relevance for the current request since the HTTP Method is '" +
+							httpMethod + "', and no request body will be sent for that method.");
+				}
+			}
+
+			Object expectedResponseType = this.determineExpectedResponseType(requestMessage);
+
+			HttpEntity<?> httpRequest = this.generateHttpRequest(requestMessage, httpMethod);
+			Map<String, ?> uriVariables = this.determineUriVariables(requestMessage);
+			UriComponentsBuilder uriComponentsBuilder = uri instanceof String
+					? UriComponentsBuilder.fromUriString((String) uri)
+					: UriComponentsBuilder.fromUri((URI) uri);
+			UriComponents uriComponents = uriComponentsBuilder.buildAndExpand(uriVariables);
+			realUri = this.encodeUri ? uriComponents.toUri() : new URI(uriComponents.toUriString());
+
+			return this.getResponse(expectedResponseType, realUri, httpMethod, httpRequest);
+		}
+		catch (MessagingException e) {
+			throw e;
+		}
+		catch (Exception e) {
+			throw new MessageHandlingException(requestMessage, "HTTP request execution failed for URI ["
+					+ (realUri == null ? uri : realUri) + "]", e);
+		}
+	}
+
+	abstract protected Object getResponse(Object expectedResponseType, URI realUri, HttpMethod httpMethod, HttpEntity<?> httpRequest);
+
+	protected Object getReply(ResponseEntity<?> httpResponse) {
+		if (this.expectReply) {
+			HttpHeaders httpHeaders = httpResponse.getHeaders();
+			Map<String, Object> headers = this.headerMapper.toHeaders(httpHeaders);
+			if (this.transferCookies) {
+				this.doConvertSetCookie(headers);
+			}
+			AbstractIntegrationMessageBuilder<?> replyBuilder = null;
+			if (httpResponse.hasBody()) {
+				Object responseBody = httpResponse.getBody();
+				replyBuilder = (responseBody instanceof Message<?>) ?
+						this.getMessageBuilderFactory().fromMessage((Message<?>) responseBody) : this.getMessageBuilderFactory().withPayload(responseBody);
+
+			}
+			else {
+				replyBuilder = this.getMessageBuilderFactory().withPayload(httpResponse);
+			}
+			replyBuilder.setHeader(org.springframework.integration.http.HttpHeaders.STATUS_CODE, httpResponse.getStatusCode());
+			return replyBuilder.copyHeaders(headers).build();
```

Comment:
Since we are here already, let's consider to include the fix for this one as well: https://jira.spring.io/browse/INT-4076 !